docs/internal/expressions.md

```diff
@@ -0,0 +1,597 @@
+
+#### Loops
+
+Loops are also implemented using `match`, but it iterates using tail recursion.
+The code generated by this emulation, despite sounding very different, is identical to a traditional iteration because of the types involved in the iteration.
+
+For flow control in loops, `continue` and `break` are types used on the `match` to continue or stop iteration by exiting at the right place (beginning or end) in the iteration process.
+
+## Methods
+
+Methods are functions that were declared on a particular type.
+Like most other languages, methods can have arbitrary code inside, including creating asynchronous behaviours that could be executed long after the method has returned.
+
+### Declaring methods
+
+Methods can have any number of arguments (including none) and can return zero or one values.
+If a method needs to return more than one value, it can do so by returning a tuple of values.
+
+The types of the arguments and return values must be known at compile time, though they do not need to be concrete (classes).
+For non-concrete cases, their implementation needs to be complete for closed-world cases (ex. account for all types in a union) but not for open-world cases (ex. interfaces), as long as they only use features that the interfaces provides.
+
+Method names can be anything that is allowed on identifiers, including numbers and symbols.
+It must not, however, be comprised of numbers only, to avoid confusing them with numeric literals.
+
+Example:
+```ts
+class MyType
+{
+  // Sugar constructor, no args, returns the type itself
+  create() : MyType;
+
+  // One argument, one return value
+  my_method(arg : MyType) : MyType;
+
+  // No argument, no return values
+  apply();
+
+  // Multiple arguments, multiple return values
+  ===(arg0 : MyYType, arg1 : U32) : (Bool, MyType, MaybeError);
+}
+```
+
+Unlike most other languages, Verona does not have _free functions_, ie. those that are not attached to a type.
+In that sense, all Verona functions are methods of some class, for example, like `U32::+(U32, U32)` implements addition of unsigned 32-bit integers for `(a + b)`.
+
+Modules are classes, so they can have functions without declaring further types, but those are just methods of the module's class type.
+So a file can have functions outside of declared types that look like free functions, but to call them, one must either pass the full name (ex. `MyModule::function()`) or `use` a module to import that module's context on your own.
+
+Example:
+```ts
+// On MyModule
+{
+  ...
+  foo() { ... }
+}
+
+// On OtherModule
+{
+  ...
+  foo() { ... }
+}
+
+// On some other module
+using "MyModule";
+bar()
+{
+  foo(); // This is fine, since only MyModule's scope has been imported here
+}
+
+// On some yet another module
+using "MyModule";
+using "OtherModule";
+bar()
+{
+  foo(); // ERROR: which one?
+  MyModule::foo(); // Fully qualified, this is fine.
+}
```

Comment:
I agree, but the example is still valid.

Also, this problem happens even without importing both directly:
```
// File1 has
use "OpenFoo"; // provides `foo`

// File2 has
use "SuperFoo"; // also provides `foo`

// File3 has
use "File1"
use "File2"
```

which one you use above may bring different interfaces for the same functionality. Code may have both on non-intersecting places and you only need one file to include other files that end up including both to break on a non-local error message.

Reply:
I'm going to hit resolve here, because it's not a topic for this doc, but I have some thoughts about versioning and how we can try to address some of this that I think would be good to discuss.